Introducing AGENTS.md instructions for the repo (#2695)
intel/scikit-learn-intelex: tests/AGENTS.md

## Test Execution

### Local Development
Core test suites:
- `pytest --verbose -s tests/` - Legacy/integration tests
- `pytest --verbose daal4py` - Native oneDAL API tests
- `pytest --verbose sklearnex` - sklearn compatibility tests
- `pytest --verbose onedal` - Low-level backend tests
- `pytest --verbose .ci/scripts/test_global_patch.py` - Global patching validation

Coverage reporting:
- `pytest --cov=onedal --cov=sklearnex --cov-config=.coveragerc --cov-branch`

### MPI/Distributed (SPMD) Testing

**Requirements:** MPI installation (Intel MPI or MPICH), mpi4py, NO_DIST!=1

**Setup MPI:**
```bash
# Option 1: Via conda (recommended)
conda install -c conda-forge impi-devel impi_rt mpi4py mpi=*=impi

# Option 2: Via pip (Intel MPI)
pip install impi-rt mpi4py --index-url https://software.repos.intel.com/python/pypi

# See doc/sources/distributed-mode.rst for detailed Intel MPI installation instructions
```

**Rebuild with MPI support:**
```bash
# Ensure NO_DIST is not set
unset NO_DIST
# First build the C++ oneDAL backend (includes SPMD libraries)
python setup.py develop
```

**Run distributed tests:**
```bash
# sklearnex SPMD tests

> mpirun -n 4 python tests/helper_mpi_tests.py pytest -k spmd --with-mpi --verbose sklearnex

Reviewer: ```suggestion
# sklearnex SPMD tests (requires additional pytest-mpi dependency, plus dpctl and dpnp as sklearnex spmd are on GPU)
mpirun -n 4 python -m pytest --pyargs sklearnex.spmd --with-mpi
```

# daal4py SPMD examples
mpirun -n 4 python tests/helper_mpi_tests.py pytest --verbose -s tests/test_daal4py_spmd_examples.py
```

> **Validates:** DBSCAN, K-Means, PCA, Linear Regression, Covariance, Random Forest (distributed algorithms)

Reviewer: ```suggestion
**Validates:** DBSCAN, K-Means, PCA, Linear Regression, Covariance, Random Forest, KNN, Logistic Regression, and Basic Statistics (distributed algorithms)
```

   - Numerical precision variations from optimized computation paths

2. **Feature Coverage Gaps**
   - sklearn features not yet implemented in oneDAL backend
   - Emerging sklearn APIs (new versions) not yet supported

3. **Platform/Environment Constraints**
   - Operating system specific behaviors
   - Build toolchain differences (compiler-specific results)

4. **Test Infrastructure Differences**
   - Exception message wording differences (same error conditions, different text)
   - Edge case handling variations (rare scenarios handled differently but correctly)

**Version-Conditional Deselection**: Tests can be deselected for specific sklearn version ranges using comparison operators in `deselected_tests.yaml`.

**Impact**: Deselected tests represent a small subset of sklearn's test suite. The vast majority of algorithms work identically to sklearn. Check `deselected_tests.yaml` for specifics and see inline comments for each deselection reason.

### Running Compatibility Tests
```bash
# Full sklearn compatibility test
pytest --verbose sklearnex

# Tests respect deselected_tests.yaml automatically
# To see what's deselected: cat deselected_tests.yaml

> ```

Reviewer: ```suggestion
Tests are configured in .circleci/
```bash
# Standard run
python .circleci/run_xpu_tests.py -q -d cpu --reduced --deselected_yml_file deselected_tests.yaml

# Run on GPU by specifying device and adding gpu deselections
python .circleci/run_xpu_tests.py -q -d gpu --reduced --gpu --deselected_yml_file deselected_tests.yaml

# Run tests on stock scikit-learn to compare results
python .circleci/run_xpu_tests.py -q --no-intel-optimized -d cpu --reduced --deselected_yml_file deselected_tests.yaml
```
```

- `pytest --verbose sklearnex` - sklearn compatibility tests
- `pytest --verbose onedal` - Low-level backend tests
- `pytest --verbose .ci/scripts/test_global_patch.py` - Global patching validation

Coverage reporting:
- `pytest --cov=onedal --cov=sklearnex --cov-config=.coveragerc --cov-branch`

### MPI/Distributed (SPMD) Testing

**Requirements:** MPI installation (Intel MPI or MPICH), mpi4py, NO_DIST!=1

**Setup MPI:**
```bash
# Option 1: Via conda (recommended)
conda install -c conda-forge impi-devel impi_rt mpi4py=*=*impi*

# Option 2: System package manager
# Debian/Ubuntu:
sudo apt-get install libopenmpi-dev openmpi-bin
pip install mpi4py

# RedHat/CentOS:
sudo yum install openmpi-devel
pip install mpi4py

# Set MPIROOT if not using conda
export MPIROOT=/path/to/mpi
```

**Rebuild with MPI support:**
```bash
# Ensure NO_DIST is not set
unset NO_DIST
python setup.py build_ext --inplace --force
```

**Run distributed tests:**
```bash
# sklearnex SPMD tests
mpirun -n 4 python tests/helper_mpi_tests.py pytest -k spmd --with-mpi --verbose sklearnex

# daal4py SPMD examples
mpirun -n 4 python tests/helper_mpi_tests.py pytest --verbose -s tests/test_daal4py_spmd_examples.py
```

> **Validates:** DBSCAN, K-Means, PCA, Linear Regression, Covariance (distributed algorithms)

Author: ```suggestion
**Validates:** DBSCAN, K-Means, PCA, Linear Regression, Covariance, Random Forest (distributed algorithms)
```

1. **Intentional differences**: Performance optimizations that produce mathematically equivalent but not identical results
2. **Unsupported features**: Features not yet implemented in oneDAL backend
3. **Implementation constraints**: Limitations in oneDAL library or SYCL/GPU execution
4. **Platform-specific issues**: Environment-dependent test failures

### Deselected Tests

**Configuration**: `deselected_tests.yaml` - See this file for current test list, counts, and specific reasons.

**Why Tests Are Deselected** (high-level categories):

1. **Intentional Implementation Differences**
   - Performance-optimized algorithms producing mathematically equivalent but not bit-identical results
   - Different internal algorithms achieving same outcomes (e.g., alternative solver selection)
   - Numerical precision variations from optimized computation paths

2. **Feature Coverage Gaps**
   - sklearn features not yet implemented in oneDAL backend
   - Emerging sklearn APIs (new versions) not yet supported
   - Experimental features (Array API, specific solvers) with partial support

3. **Platform/Environment Constraints**

>    - GPU/SYCL execution limitations

Author: ```suggestion
```

   - Operating system specific behaviors
   - Build toolchain differences (compiler-specific results)

4. **Test Infrastructure Differences**
   - Exception message wording differences (same error conditions, different text)
   - Parameter validation order differences (both catch invalid input, different sequence)
   - Edge case handling variations (rare scenarios handled differently but correctly)

**Version-Conditional Deselection**: Tests can be deselected for specific sklearn version ranges using comparison operators in `deselected_tests.yaml`.

**Impact**: Deselected tests represent a small subset of sklearn's test suite. The vast majority of algorithms work identically to sklearn. Check `deselected_tests.yaml` for specifics and see inline comments for each deselection reason.

### Running Compatibility Tests
```bash
# Full sklearn compatibility test

> pytest --verbose --pyargs sklearnex

Reviewer 2: ```suggestion
pytest --verbose sklearnex
```

# Tests respect deselected_tests.yaml automatically
# To see what's deselected: cat deselected_tests.yaml
```

## Key Testing Patterns
- Use `np.testing.assert_allclose(atol=1e-05)` for numerical validation
- Configure timeouts based on algorithm complexity (default 170s, complex up to 480s)

> - Handle missing dependencies with `skipTest()`

Reviewer 2: ```suggestion
- Handle missing dependencies with `pytest.skip()`
```
